### PR TITLE
Allow IPv6 datagram protocol addresses

### LINKED ISSUE #15169: asyncio: `DatagramProtocol.datagram_received` has incorrect annotation for `addr` parameter
`DatagramProtocol.datagram_received` is annotated as:

```python
def datagram_received(self, data: bytes, addr: tuple[str | Any, int]) -> None: ...
```

However, [when using IPv6](https://docs.python.org/3/library/socket.html#socket-families), the type of `addr` will be `tuple[str, int, int, int]` (`(host, port, flowinfo, scope_id)`).

### PR DESCRIPTION
**Pull request by popsiclelmlm**
Summary
- Widen `asyncio.DatagramProtocol.datagram_received`'s `addr` parameter to use `socket._RetAddress`.
- Add regression coverage for IPv4 and IPv6 datagram protocol overrides.

Reproduction
- Issue #15169 reports that IPv6 datagram endpoints can pass `(host, port, flowinfo, scope_id)`, but the stub only allowed a two-item address tuple.

Root cause
- `DatagramProtocol.datagram_received` used a two-item tuple type as a practical approximation even though datagram addresses vary by socket family.
- `socket.recvfrom` already models returned addresses as `_RetAddress`, which is broad enough for IPv4, IPv6, and non-IP socket families.

Changes
- Import `socket._RetAddress` in `stdlib/asyncio/protocols.pyi`.
- Use `_RetAddress` for `DatagramProtocol.datagram_received`.
- Add `stdlib/@tests/test_cases/asyncio/check_datagram_protocol.py` with IPv4 and IPv6 override examples.

Tests
- `PATH="$PWD/.venv/bin:$PATH" pre-commit run --files stdlib/asyncio/protocols.pyi stdlib/@tests/test_cases/asyncio/check_datagram_protocol.py`
- `PATH="$PWD/.venv/bin:$PATH" .venv/bin/python tests/regr_test.py stdlib --platform darwin -p 3.10`
- `PATH="$PWD/.venv/bin:$PATH" npx pyright@1.1.410 stdlib/@tests/test_cases --pythonversion 3.10 -p pyrightconfig.testcases.json`
- `PATH="$PWD/.venv/bin:$PATH" npx pyright@1.1.410 stdlib/asyncio --pythonversion 3.10 -p pyrightconfig.stricter.json`
- `git diff --check`

Screenshots/Logs
- Not applicable; stdlib stub/test-only change.
- Note: `tests/runtests.py stdlib/asyncio` also ran locally. Its pyright/pre-commit phases passed, but stdlib stubtest failed on this local Python 3.14 environment due unrelated runtime availability/difference issues such as missing `_tkinter`, missing `_curses.BUTTON5_*`, and annotationlib differences.

AI disclosure
- Prepared with assistance from Codex.

Fixes #15169

### CODE REVIEW

File: stdlib/@tests/test_cases/asyncio/check_datagram_protocol.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+import asyncio
+
+
+class IPv4DatagramProtocol(asyncio.DatagramProtocol):
+    def datagram_received(self, data: bytes, addr: tuple[str, int]) -> None: ...
+
+
+class IPv6DatagramProtocol(asyncio.DatagramProtocol):
+    def datagram_received(self, data: bytes, addr: tuple[str, int, int, int]) -> None: ...
```
Comment (srittau) [inline stdlib/@tests/test_cases/asyncio/check_datagram_protocol.py:11]: For performance and other reasons we only use tests in extraordinary or edge cases. These tests basically test whether type checkers process `Any` correctly. We don't need them.

File: stdlib/asyncio/protocols.pyi [hunk 2]
```diff
@@ -27,11 +27,10 @@ class BufferedProtocol(BaseProtocol):
...
     def connection_made(self, transport: transports.DatagramTransport) -> None: ...  # type: ignore[override]
-    # addr can be a tuple[int, int] for some unusual protocols like socket.AF_NETLINK.
-    # Use tuple[str | Any, int] to not cause typechecking issues on most usual cases.
-    # This could be improved by using tuple[AnyOf[str, int], int] if the AnyOf feature is accepted.
-    # See https://github.com/python/typing/issues/566
-    def datagram_received(self, data: bytes, addr: tuple[str | Any, int]) -> None: ...
+    # addr varies by socket family and can be a tuple[str, int, int, int] for IPv6,
+    # tuple[int, int] for some unusual protocols like socket.AF_NETLINK, or other
+    # shapes. Keep this aligned with socket.recvfrom's return address type.
+    def datagram_received(self, data: bytes, addr: _RetAddress) -> None: ...
```
Comment (srittau) [inline stdlib/asyncio/protocols.pyi:33]: Using `Any` here makes sense to me, as this is hard to type correctly, but we should just use it directly than obscuring it by going through the `_RetAddress` alias:


```suggestion
    # addr varies by socket family and can be a tuple[str, int, int, int] for IPv6,
    # tuple[int, int] for some unusual protocols like socket.AF_NETLINK, or other
    # shapes.
    def datagram_received(self, data: bytes, addr: Any) -> None: ...
```

**Review by srittau — CHANGES_REQUESTED** [review]
